File: 1195. Fizz Buzz Multithreaded/main.py
import threading
from typing import Callable


class FizzBuzz:
    def hasFinished(self) -> bool:
        return self.cur > self.n

    def isFizz(self) -> bool:
        return (self.cur % 3 == 0) and (self.cur % 5 != 0)

    def isBuzz(self) -> bool:
        return (self.cur % 3 != 0) and (self.cur % 5 == 0)

    def isFizzBuzz(self) -> bool:
        return (self.cur % 3 == 0) and (self.cur % 5 == 0)

    def isNumber(self) -> bool:
        return (self.cur % 3 != 0) and (self.cur % 5 != 0)

    def __init__(self, n: int):
        self.n = n
        self.cur = 1
        self.condition = threading.Condition()

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: "Callable[[], None]") -> None:
        while True:
            with self.condition:
                while not self.hasFinished() and not self.isFizz():
                    self.condition.wait()

                if self.hasFinished():
                    return

                printFizz()
                self.cur += 1

                self.condition.notify_all()

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: "Callable[[], None]") -> None:
        while True:
            with self.condition:
                while not self.hasFinished() and not self.isBuzz():
                    self.condition.wait()

                if self.hasFinished():
                    return

                printBuzz()
                self.cur += 1

                self.condition.notify_all()

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: "Callable[[], None]") -> None:
        while True:
            with self.condition:
                while not self.hasFinished() and not self.isFizzBuzz():
                    self.condition.wait()

                if self.hasFinished():
                    return

                printFizzBuzz()
                self.cur += 1

                self.condition.notify_all()

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: "Callable[[int], None]") -> None:
        while True:
            with self.condition:
                while not self.hasFinished() and not self.isNumber():
                    self.condition.wait()

                if self.hasFinished():
                    return

                printNumber(self.cur)
                self.cur += 1

                self.condition.notify_all()


class FizzBuzz:
    def hasFinished(self) -> bool:
        return self.cur > self.n

    def isFizz(self) -> bool:
        return (self.cur % 3 == 0) and (self.cur % 5 != 0)

    def isBuzz(self) -> bool:
        return (self.cur % 3 != 0) and (self.cur % 5 == 0)

    def isFizzBuzz(self) -> bool:
        return (self.cur % 3 == 0) and (self.cur % 5 == 0)

    def isNumber(self) -> bool:
        return (self.cur % 3 != 0) and (self.cur % 5 != 0)

    def __init__(self, n: int):
        self.n = n
        self.cur = 1
        self.lock = threading.RLock()

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: "Callable[[], None]") -> None:
        while not self.hasFinished():
            with self.lock:
                if not self.hasFinished() and self.isFizz():
                    printFizz()
                    self.cur += 1

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: "Callable[[], None]") -> None:
        while not self.hasFinished():
            with self.lock:
                if not self.hasFinished() and self.isBuzz():
                    printBuzz()
                    self.cur += 1

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: "Callable[[], None]") -> None:
        while not self.hasFinished():
            with self.lock:
                if not self.hasFinished() and self.isFizzBuzz():
                    printFizzBuzz()
                    self.cur += 1

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: "Callable[[int], None]") -> None:
        while not self.hasFinished():
            with self.lock:
                if not self.hasFinished() and self.isNumber():
                    printNumber(self.cur)
                    self.cur += 1

File: 1195. Fizz Buzz Multithreaded/main.c
#include <pthread.h>
#include <stdbool.h>
#include <stdio.h>

typedef struct {
    int n;
    int cur;
    pthread_mutex_t mutex;
    pthread_cond_t cond;
} FizzBuzz;

FizzBuzz *fizzBuzzCreate(int n) {
    FizzBuzz *obj = (FizzBuzz *)malloc(sizeof(FizzBuzz));
    if (obj == NULL) {
        return NULL;
    }

    obj->n = n;
    obj->cur = 1;
    pthread_mutex_init(&obj->mutex, NULL);
    pthread_cond_init(&obj->cond, NULL);
    return obj;
}

static bool hasFinished(FizzBuzz *obj) {
    return obj->cur > obj->n;
}

static bool isFizz(FizzBuzz *obj) {
    return (obj->cur % 3 == 0) && (obj->cur % 5 != 0);
}

static bool isBuzz(FizzBuzz *obj) {
    return (obj->cur % 3 != 0) && (obj->cur % 5 == 0);
}

static bool isFizzBuzz(FizzBuzz *obj) {
    return (obj->cur % 3 == 0) && (obj->cur % 5 == 0);
}

static bool isNumber(FizzBuzz *obj) {
    return (obj->cur % 3 != 0) && (obj->cur % 5 != 0);
}

// printFizz() outputs "fizz".
void fizz(FizzBuzz *obj) {
    bool looping = true;

    while (looping) {
        pthread_mutex_lock(&obj->mutex);
        while (!hasFinished(obj) && !isFizz(obj)) {
            pthread_cond_wait(&obj->cond, &obj->mutex);
        }

        if (hasFinished(obj)) {
            looping = false;
        } else {
            printFizz();
            obj->cur++;
        }
        pthread_cond_broadcast(&obj->cond);
        pthread_mutex_unlock(&obj->mutex);
    }
}

// printBuzz() outputs "buzz".
void buzz(FizzBuzz *obj) {
    bool looping = true;

    while (looping) {
        pthread_mutex_lock(&obj->mutex);
        while (!hasFinished(obj) && !isBuzz(obj)) {
            pthread_cond_wait(&obj->cond, &obj->mutex);
        }

        if (hasFinished(obj)) {
            looping = false;
        } else {
            printBuzz();
            obj->cur++;
        }
        pthread_cond_broadcast(&obj->cond);
        pthread_mutex_unlock(&obj->mutex);
    }
}

// printFizzBuzz() outputs "fizzbuzz".
void fizzbuzz(FizzBuzz *obj) {
    bool looping = true;

    while (looping) {
        pthread_mutex_lock(&obj->mutex);
        while (!hasFinished(obj) && !isFizzBuzz(obj)) {
            pthread_cond_wait(&obj->cond, &obj->mutex);
        }

        if (hasFinished(obj)) {
            looping = false;
        } else {
            printFizzBuzz();
            obj->cur++;
        }
        pthread_cond_broadcast(&obj->cond);
        pthread_mutex_unlock(&obj->mutex);
    }
}

// You may call global function `void printNumber(int x)`
// to output "x", where x is an integer.
void number(FizzBuzz *obj) {
    bool looping = true;

    while (looping) {
        pthread_mutex_lock(&obj->mutex);
        while (!hasFinished(obj) && !isNumber(obj)) {
            pthread_cond_wait(&obj->cond, &obj->mutex);
        }

        if (hasFinished(obj)) {
            looping = false;
        } else {
            printNumber(obj->cur);
            obj->cur++;
        }
        pthread_cond_broadcast(&obj->cond);
        pthread_mutex_unlock(&obj->mutex);
    }
}

void fizzBuzzFree(FizzBuzz *obj) {
    if (obj == NULL) {
        return;
    }

    pthread_mutex_destroy(&obj->mutex);
    pthread_cond_destroy(&obj->cond);
    free(obj);
}